config/v1/types_apiserver.go

```diff
@@ -234,6 +240,36 @@ const (
 	EncryptionTypeKMS EncryptionType = "KMS"
 )
 
+// +union
+// +kubebuilder:validation:XValidation:rule="self.mode == 'CustomDeployment' ? has(self.customDeployment) : !has(self.customDeployment)",message="customDeployment is required when mode is CustomDeployment and forbidden otherwise"
+type HTTP01ChallengeProxySpec struct {
+	// mode controls whether the HTTP01 challenge proxy is active and how it should be deployed.
+	// DefaultDeployment enables the proxy with default configuration.
+	// CustomDeployment enables the proxy with user-specified configuration.
+	// +kubebuilder:validation:Enum=DefaultDeployment;CustomDeployment
+	// +kubebuilder:default=DefaultDeployment
+	// +optional
+	// +unionDiscriminator
+	Mode string `json:"mode,omitempty"`
+
+	// customDeployment contains configuration options when mode is CustomDeployment.
+	// This field is only valid when mode is CustomDeployment.
+	// +optional
+	// +unionMember
+	CustomDeployment HTTP01ChallengeProxyCustomDeploymentSpec `json:"customDeployment,omitzero,omitempty"`
+}
+
+// +kubebuilder:validation:MinProperties=1
```

Comment:
Not strictly needed since you have a required field, the zero value `{}` would fail validation on the required field not being present, and `interalPort: 0` would also fail validation because `0` is not valid.

You can leave this though if you prefer to keep it